core/unit_test/TestDeepCopy.hpp

```diff
@@ -210,5 +211,147 @@ TEST(TEST_CATEGORY, deep_copy_alignment) {
                        Kokkos::HostSpace>::run_test(100000);
   }
 }
+#endif
+
+namespace Impl {
+template <class Scalar1, class Scalar2, class Layout1, class Layout2>
+struct TestDeepCopyScalarConversion {
+  struct TagFill {};
+  struct TagCompare {};
+
+  using view_type_s1_1d = Kokkos::View<Scalar1*, Layout1, TEST_EXECSPACE>;
+  using view_type_s2_1d = Kokkos::View<Scalar2*, Layout2, TEST_EXECSPACE>;
+  using view_type_s1_2d = Kokkos::View<Scalar1**, Layout1, TEST_EXECSPACE>;
+  using view_type_s2_2d = Kokkos::View<Scalar2**, Layout2, TEST_EXECSPACE>;
+
+  using base_layout1 = typename std::conditional<
+      std::is_same<Layout1, Kokkos::LayoutStride>::value, Kokkos::LayoutLeft,
+      Layout1>::type;
+  using base_layout2 = typename std::conditional<
+      std::is_same<Layout2, Kokkos::LayoutStride>::value, Kokkos::LayoutLeft,
+      Layout2>::type;
+
+  using base_type_s1_1d = Kokkos::View<Scalar1*, base_layout1, TEST_EXECSPACE>;
+  using base_type_s2_1d = Kokkos::View<Scalar2*, base_layout2, TEST_EXECSPACE>;
+  using base_type_s1_2d = Kokkos::View<Scalar1**, base_layout1, TEST_EXECSPACE>;
+  using base_type_s2_2d = Kokkos::View<Scalar2**, base_layout2, TEST_EXECSPACE>;
+
+  view_type_s1_1d view_s1_1d;
+  view_type_s2_1d view_s2_1d;
+  view_type_s1_2d view_s1_2d;
+  view_type_s2_2d view_s2_2d;
+
+  Kokkos::View<int64_t, TEST_EXECSPACE> error_count;
+
+  void create_views(int64_t N0, int64_t N1) {
+    base_type_s1_1d b_s1_1d("TestDeepCopyConversion::b_s1_1d", N0);
+    base_type_s2_1d b_s2_1d("TestDeepCopyConversion::b_s2_1d", N0);
+    base_type_s1_2d b_s1_2d("TestDeepCopyConversion::b_s1_2d", N0, N1);
+    base_type_s2_2d b_s2_2d("TestDeepCopyConversion::b_s2_2d", N0, N1);
+
+    view_s1_1d = view_type_s1_1d(b_s1_1d, Kokkos::ALL);
+    view_s2_1d = view_type_s2_1d(b_s2_1d, Kokkos::ALL);
+    view_s1_2d = view_type_s1_2d(b_s1_2d, Kokkos::ALL, Kokkos::ALL);
+    view_s2_2d = view_type_s2_2d(b_s2_2d, Kokkos::ALL, Kokkos::ALL);
+
+    error_count = Kokkos::View<int64_t, TEST_EXECSPACE>(
+        "TestDeepCopyConversion::error_count");
+  }
+
+  KOKKOS_FUNCTION
+  void operator()(TagFill, const int64_t i) const {
+    view_s2_1d(i) = static_cast<Scalar2>(i + 1);
+    for (int64_t j = 0; j < static_cast<int64_t>(view_s2_2d.extent(1)); j++)
+      view_s2_2d(i, j) = static_cast<Scalar2>((i + 1) * 1000 + j + 1);
+  }
+
+  KOKKOS_FUNCTION
+  void operator()(TagCompare, const int64_t i) const {
+    int64_t errors = 0;
+    if (view_s1_1d(i) != static_cast<Scalar1>(static_cast<Scalar2>(i + 1)))
+      errors++;
+    for (int64_t j = 0; j < static_cast<int64_t>(view_s1_2d.extent(1)); j++) {
+      if (view_s1_2d(i, j) !=
+          static_cast<Scalar1>(static_cast<Scalar2>((i + 1) * 1000 + j + 1)))
+        errors++;
+    }
+    if (errors > 0) Kokkos::atomic_add(&error_count(), errors);
```

Comment:
Looks like there is a problem with the atomic_add with HIP, `errors` has the correct value. We haven't really worked on the `atomic_add` in HIP. So it's not surprising that it doesn't work.

Reply:
Ok gonna disable the test for HIP for now.